### DISCUSSION TITLE
The stereo information is removed after reaction.RunReactants

### DISCUSSION THREAD

**Discussion by shuan4638** [rdkit/rdkit#4096] [Q&A] [answered]
I tried to run the reaction as following
```
from rdkit import Chem
from rdkit.Chem import rdChemReactions
reactant1 = Chem.MolFromSmiles('CCOC(=O)/C=C/I')
reactant2 = Chem.MolFromSmiles('O=Cc1ccc2nc(-c3cccs3)oc2c1')
template = 'I-[C;H1;+0:1].[C;H1;+0:2]=[O;H0;+0:3]>>[C;H1;+0:1]-[C;H1;+0:2]-[O;H1;+0:3]'
reaction = rdChemReactions.ReactionFromSmarts(template)
reactants = reaction.RunReactants([reactant1, reactant2])
print (Chem.MolToSmiles(reactants[0][0]))
```
I expected the output will remain the CIS/TRANS conformation as it was in `reactant1`, such
```
'CCOC(=O)/C=C/C(O)c1ccc2nc(-c3cccs3)oc2c1'
```
However, I got 
```
'CCOC(=O)C=CC(O)c1ccc2nc(-c3cccs3)oc2c1'
```

Why does this happen and how should I solve this issue?

**Comment by greglandrum** [comment, marked as answer]
@shuan4638 : this is a tricky one.
The reason the bond stereo is being lost is that you're breaking and reforming one of the bonds around the double bond. The RDKit doesn't have any way of knowing that you want to preserve the double bond stereo in this situation. 

If you know that the reaction is going to take place at a double bond and preserve the stereochemistry (as in your example) you need to capture that extra information in your reaction defintion:
```
from rdkit import Chem
from rdkit.Chem import rdChemReactions
reactant1 = Chem.MolFromSmiles('CCOC(=O)/C=C/I')
reactant2 = Chem.MolFromSmiles('O=Cc1ccc2nc(-c3cccs3)oc2c1')
template = 'I/[C;H1;+0:1]=[C:4]/[*:5].[C;H1;+0:2]=[O;H0;+0:3]>>[*:5]/[C:4]=[C;H1;+0:1]/[C;H1;+0:2]-[O;H1;+0:3]'
reaction = rdChemReactions.ReactionFromSmarts(template)
reactants = reaction.RunReactants([reactant1, reactant2])
print (Chem.MolToSmiles(reactants[0][0]))
```
The output is the expected:
```
CCOC(=O)/C=C/C(O)c1ccc2nc(-c3cccs3)oc2c1
```

**Reply by cottonkiet** [reply]
This solution works when reactant1 is trans isomer. However, when reactant1 is cis isomer
`reactant1 = Chem.MolFromSmiles('CCOC(=O)/C=C\I')`
the output is still trans:
`CCOC(=O)/C=C/C(O)c1ccc2nc(-c3cccs3)oc2c1`
There are two questions. why is the cis reactant even recognised in the reaction querry, and even if it is recognised, why the product is still trans? Thanks a lot!

**Comment by bp-kelley** [comment]
I'm hazarding a guess that you might not have changed the stereo on *both* sides of the reaction:

```
template = 'I/[C;H1;+0:1]=[C:4]\[*:5].[C;H1;+0:2]=[O;H0;+0:3]>>[*:5]\[C:4]=[C;H1;+0:1]/[C;H1;+0:2]-[O;H1;+0:3]'
```

But what you really might find annoying is that chirality is NOT used during the reactant matching step if this is important to you, you need to filter them out first doing something like the following (this is a leftover from the days where the reaction chirality wasn't handled very well)

```
from rdkit import Chem
from rdkit.Chem import rdChemReactions
reactant1 = Chem.MolFromSmiles('CCOC(=O)/C=C\I')
reactant2 = Chem.MolFromSmiles('O=Cc1ccc2nc(-c3cccs3)oc2c1')
template = 'I/[C;H1;+0:1]=[C:4]\[*:5].[C;H1;+0:2]=[O;H0;+0:3]>>[*:5]\[C:4]=[C;H1;+0:1]/[C;H1;+0:2]-[O;H1;+0:3]'
reaction = rdChemReactions.ReactionFromSmarts(template)
# filter out bad chirality matches to reactants
pats = list(reaction.GetReactants())
if reactant1.HasSubstructMatch(pats[0], useChirality=True) and reactant2.HasSubstructMatch(pats[1], useChirality=True):
    reactants = reaction.RunReactants([reactant1, reactant2])
    print (Chem.MolToSmiles(reactants[0][0]))
```

This outputs the expected

```
CCOC(=O)/C=C\C(O)c1ccc2nc(-c3cccs3)oc2c1
```

**Reply by cottonkiet** [reply]
Thank you a lot! this works. so I would create two separate schemes, one for trans and one for cis. 
However, this does not work if the reactant1 has two reaction centers. for example.
reactant1 = `Chem.MolFromSmiles('I\C=C\C(=O)OCCCCOC(=O)\C=C/I')`
Is there any way to create the filter inside RunReactants function or use mapping so that it knows which double bond the reaction is changing? 
Thanks!

**Reply by bp-kelley** [reply]
If you have multiple reacting centers, you are going to get multiple products:

```
    reactant1 = Chem.MolFromSmiles('I\C=C\C(=O)OCCCCOC(=O)\C=C/I')
    ...
    reactants = reaction.RunReactants([reactant1, reactant2])
    for product in reactants:
        print(Chem.MolToSmiles(product[0]))
```

will output

```
O=C(/C=C\I)OCCCCOC(=O)/C=C\C(O)c1ccc2nc(-c3cccs3)oc2c1
O=C(/C=C\C(O)c1ccc2nc(-c3cccs3)oc2c1)OCCCCOC(=O)/C=C/I
```

**Reply by cottonkiet** [reply]
Yes I will get multiple products. but the product 1 in your output does not preserve the trans double bond as in the original reactant. What I would like to do is retention of stereochemistry when doing the reaction. but the filter will pass if another moiety matches the pattern, which is not the moiety that actually undergo the reaction.

**Comment by bp-kelley** [comment]
In this case, just don't specify the cis/trans in the output product and the reaction should copy it over:

```
template = 'I/[C;H1;+0:1]=[C:4]\[*:5].[C;H1;+0:2]=[O;H0;+0:3]>>[*:5][C:4]=[C;H1;+0:1][C;H1;+0:2]-[O;H1;+0:3]'
```

output
```
O=C(C=CC(O)c1ccc2nc(-c3cccs3)oc2c1)OCCCCOC(=O)/C=C\I
O=C(C=CC(O)c1ccc2nc(-c3cccs3)oc2c1)OCCCCOC(=O)/C=C/I
```

**Reply by cottonkiet** [reply]
it seems that the stereo information is removed if we does not specify the cis/trans in the output product. 
my idea output is something like this
```
O=C(/C=C\I)OCCCCOC(=O)/C=C/C(O)c1ccc2nc(-c3cccs3)oc2c1
O=C(/C=C\C(O)c1ccc2nc(-c3cccs3)oc2c1)OCCCCOC(=O)/C=C/I
```

**Comment by bp-kelley** [comment]
What rdkit version are you running?

```
from rdkit import Chem
from rdkit.Chem import rdChemReactions
import rdkit
print("version:", rdkit.__version__)
reactant1 = Chem.MolFromSmiles('I\C=C\C(=O)OCCCCOC(=O)\C=C/I')
reactant2 = Chem.MolFromSmiles('O=Cc1ccc2nc(-c3cccs3)oc2c1')
template = 'I[C;H1;+0:1]=[C:4][*:5].[C;H1;+0:2]=[O;H0;+0:3]>>[*:5][C:4]=[C;H1;+0:1][C;H1;+0:2]-[O;H1;+0:3]'
reaction = rdChemReactions.ReactionFromSmarts(template)
# filter out bad chirality matches to reactants
pats = list(reaction.GetReactants())
if reactant1.HasSubstructMatch(pats[0], useChirality=True) and reactant2.HasSubstructMatch(pats[1], useChirality=True):
    reactants = reaction.RunReactants([reactant1, reactant2])
    for product in reactants:
        print(Chem.MolToSmiles(product[0]))
```

output

```
version: 2021.09.4
O=C(C=CC(O)c1ccc2nc(-c3cccs3)oc2c1)OCCCCOC(=O)/C=C\I
O=C(C=CC(O)c1ccc2nc(-c3cccs3)oc2c1)OCCCCOC(=O)/C=C/I
```

**Reply by cottonkiet** [reply]
my version is 2021.09.5
I get the same result as yours.
however, the output has removed stereochemistry, which is not I want.

what I want is
```
O=C(/C=C\I)OCCCCOC(=O)/C=C/C(O)c1ccc2nc(-c3cccs3)oc2c1
O=C(/C=C\C(O)c1ccc2nc(-c3cccs3)oc2c1)OCCCCOC(=O)/C=C/I
```

**Reply by cottonkiet** [reply]
what I mean is your output removed stereochemistry of one double bond.

**Reply by bp-kelley** [reply]
Since you are removing an iodine, you need to map it to something to preserve cis/trans

I'm not entirely sure this is what you want, but try mapping it to an explicit hydrogen ([I:6] -> [H:6])

```
template = '[I:6][C;H1;+0:1]=[C:4].[C;H1;+0:2]=[O;H0;+0:3]>>[C:4]=[C:1]([H:6])[C:2]-[O:3]'
reactant1 = Chem.MolFromSmiles('I\C=C\C(=O)OCCCCOC(=O)\C=C/I')
reactant2 = Chem.MolFromSmiles('O=Cc1ccc2nc(-c3cccs3)oc2c1')
reaction = rdChemReactions.ReactionFromSmarts(template)
# filter out bad chirality matches to reactants
pats = list(reaction.GetReactants())
if reactant1.HasSubstructMatch(pats[0], useChirality=True) and reactant2.HasSubstructMatch(pats[1], useChirality=True):
    reactants = reaction.RunReactants([reactant1, reactant2])
    for product in reactants:
        prod = product[0]
        Chem.SanitizeMol(prod)
        print(Chem.MolToSmiles(prod))
```

I also cleaned up the [*:5] as this was confusing the stereo here and removed the query atoms in the product as this was failing sanitization.

```
[H]/C(=C/C(=O)OCCCCOC(=O)/C=C\I)C(O)c1ccc2nc(-c3cccs3)oc2c1
[H]/C(=C\C(=O)OCCCCOC(=O)/C=C/I)C(O)c1ccc2nc(-c3cccs3)oc2c1
```

I *think* this is what you want.

**Reply by cottonkiet** [reply]
Thanks so much. This looks a nice way. However, this actually reversed the stereo chemistry. lol.

**Reply by cottonkiet** [reply]
If this make it clear. What I actually want is to preserve the stereo chemistry and create two separate reactions.  the first reaction is to perform this reaction when the bond to be transformed is trans. and the second reaction is to perform this reaction when the bond to be transformed is cis.  I am doing this because the two reactions have different reaction rates.
For example, when doing the first reaction (trans):
if the reactant is 
`reactant1 = Chem.MolFromSmiles('CCOC(=O)/C=C/I')`
the reaction does not pass
if the reactant is 
`reactant1 = Chem.MolFromSmiles('CCOC(=O)/C=C\I')`
the output is 
`[H]\C(=C\C(=O)OCC)C(O)c1ccc2nc(-c3cccs3)oc2c1`
if the reactant is
`reactant1 = Chem.MolFromSmiles('I\C=C\C(=O)OCCCCOC(=O)\C=C/I')` (two double bonds but only one trans)
the output is
`[H]\C(=C\C(=O)OCCCCOC(=O)\C=C\I)C(O)c1ccc2nc(oc2c1)-c1cccs1` (not two products here)

**Reply by bp-kelley** [reply]
You will get two products because the matching routine inside RunReactants currently doesn't check for stereo when matching the template.   This might be a good feature request now that the stereo handling after the matching works better than it used to.

**Comment by cottonkiet** [comment]
It will be definitely nice that RunReactants can have an option to check for stereo when matching the template!

**Comment by cottonkiet** [comment]
Just for discussion to answer @shuan4638 's initial question. Based on @bp-kelley 's idea , this should preserve the stereo chemistry (if not differentiate trans or cis bond). 
The first step is to add explicit H to facilitate the reaction. 
```
reactant1 = Chem.AddHs(reactant1)
template = '[I:6][C;H1;+0:1]([#1:7])=[C:4].[C;H1;+0:2]=[O;H0;+0:3]>>[C:4]=[C:1]([#1:7])[C:2]-[O:3]'
```
